### ISSUE TITLE
Apple CI Executors 

### ISSUE THREAD

**Issue by maxammann** [maplibre/maplibre#142]
TL;DR: I have two pain points right now:
* Slow Apple builds on Github Actions
* Inability to run rendering tests on Apple

---

MapLibre is a cross-platform map renderer:

* `maplibre-gl-js` works on Firefox, Chrome and Safari (on Android, iOS, Windows, macOS and Linux)
* `maplibre-gl-native` works on Android, iOS, Windows, macOS and Linux
* (`maplibre-rs` works on all Browsers and OS from above)

## Problem: Compiling for Apple

In order to compile for Apple we can use Github Actions (free). That only supports x86_64 CPUs so far. Current executors in Github CI are rather slow sadly.

### Solution

* **Cross-compile from an x86 machine to an arm64 (slow)**
* Use a paid CI service other than Github or self-host a Github Runner a real M1 (AWS infrastructure).
* Switch to another CI service for Apple builds like "XCode Cloud"

## Problem: Rendering Tests

In order to test rendering we need either real GPUs or simulated GPUs. For simulated GPUs we can use the Mesa drivers on Linux. On Windows similar libraries are available for DirectX. 
On Apple (iOS/macOS) sadly the world looks different.

In order to reliably test rendering we need CI environments which have an Apple GPU available. Historically OpenGL is supported on Apple. Nowadays Metal is state-of-the-art.

In order to run rendering test e.g. on an iPhone using Metal or WebGL in Safari we need real hardware. Note that WebGL on Safari (macOS) can either use OpenGL or Metal in the background.

### Solution

* Rent real Apple hardware 
* Use something like Browserstack (https://www.browserstack.com/).
* Switch to another CI service for Apple builds like "XCode Cloud"

## Conclusion

I think we definitely need some way to execute rendering tests and build for apple architectures.
Right now we use this for the `maplibre-gl-native` project. Though the other projects actually also need to create apple builds.

Note that compiling for Apple devices is probably possible from Github CI already through cross-compilation. Running rendering tests is not possible right now.

Also in general speeding up the compilation in Github Actions could be beneficial.

**Comment by maxammann** [comment]
I'd propose to make the current Apple M1 Github CI runner available to all projects. That would require setting up Virtual Machines on macOS so we do not cross-pollute environments.

Sadly using a custom runner on Github CI does not bring in the same environment like in the ordinary cloud runners. I do not have a solution to this.

**Comment by birkskyum** [comment]
> I'd propose to make the current Apple M1 Github CI runner available to all projects. That would require setting up Virtual Machines on macOS so we do not cross-pollute environments.

I believe that is what we have now. 3 x m1 macos runners on aws.

**Comment by maxammann** [comment]
> I believe that is what we have now. 3 x m1 macos runners on aws connected to the maplibre org. There are ongoing maintenance work keeping them stable that we have to figure out how to distribute better though.

We have 3x physical machines running? :O
I guess we could safe cost by running only one physical machine and then use VMs. 
If you mean we have 3 runners on one machine then we should probably separate them.

**Comment by birkskyum** [comment]
We have 3 physical machines running

**Comment by birkskyum** [comment]
one runner on each

**Comment by maxammann** [comment]
Alright then let's document our progress in this issue towards stable custom apple runners.

For what are you using the apple runners in the native project? Only compiling?

**Comment by birkskyum** [comment]
And testing

**Comment by wipfli** [comment]
@louwers can you review this issue? I think the state of the runners has changed since you started.

**Comment by louwers** [comment]
We now have our own macOS runners, which consist of three hosted M1Mac Mini on AWS.

They are on macOS Monterey 12.5.1 as of writing this.

Some manual set up was done (e.g. Xcode installation), but future setup should be documented and if possible automated over at:

https://github.com/maplibre/maplibre-ci-runners

You can make pull requests of issues in that repo if you want to make changes to the macOS build setup.
